pep-0577.rst

```diff
@@ -0,0 +1,652 @@
+  binding or declaration can be found in that scope, then ``TargetNameError``
+  will be raised at compile time (rather than creating a new binding within
+  the scoped expression).
+* if the containing block scope is a class scope, than ``TargetNameError`` will
+  always be raised, with a dedicated message indicating that combining class
+  scopes with augmented assignments in scoped expressions is not currently
+  permitted.
+* if a name is declared as a formal parameter (lambda expressions), or as an
+  iteration variable (generator expressions, comprehensions), then that name
+  is considered local to that scoped expression, and attempting to use it as
+  the target of an augmented assignment operation in that scope, or any nested
+  scoped expression, will raise ``TargetNameError`` (this is a restriction that
+  could potentially be lifted later, but is being proposed for now to simplify
+  the initial set of compile time and runtime semantics that needs to be
+  covered in the language reference and handled by the compiler and interpreter)
+
+For example, the following code would work as shown::
+
+    >>> global_target = 0
+    >>> incr_global_target = lambda: global_target += 1
+    >>> incr_global_target()
+    1
+    >>> incr_global_target()
+    2
+    >>> global_target
+    2
+    >>> def cumulative_sums(data, start=0)
+    ...    total = start
+    ...    yield from (total += value for value in data)
+    ...    return total
+    ...
+    >>> print(list(cumulative_sums(range(5))))
+    [0, 1, 3, 6, 10]
+
+While the following examples would all raise ``TargetNameError``::
+
+    class C:
+        cls_target = 0
+        incr_cls_target = lambda: cls_target += 1 # Error due to class scope
+
+    def missing_target():
+        incr_x = lambda: x += 1 # Error due to missing target "x"
+
+    def late_target():
+        incr_x = lambda: x += 1 # Error due to "x" being declared after use
+        x = 1
+
+    lambda arg: arg += 1 # Error due to attempt to target formal parameter
+
+    [x += 1 for x in data] # Error due to attempt to target iteration variable
+
+
+As augmented assignments currently can't appear inside scoped expressions, the
+above compile time name resolution exceptions would be included as part of the
+initial implementation rather than needing to be phased in as a potentially
+backwards incompatible change.
+
+
+Promoting nonlocal references to global references
+--------------------------------------------------
+
+As part of the above changes, all ``nonlocal NAME`` declarations (including
+the implicit ones added for augmented assignment targets in scoped expressions
+at function scope) will be changed to take explicit ``global NAME`` declarations
+into account, such that the affected name is considered ``global`` in the inner
+scope as well. For example, the following code would work by binding ``x`` in
+the global scope instead of raising ``SyntaxError`` as it does today::
```

Comment:
Why? I can see this for *implicit* nonlocal but not for explicit.

Reply:
I was thinking I needed it for consistency - I wanted to be able to explain the way `lambda: x += 1` works as always implying `global x` at module scope and `nonlocal x` at function scope.

Technically I don't *have* to define or implement it that way - I could define it as the scope declaration in the function scope getting copied down into the implicit nested scope, such that the expansion included either `nonlocal` or `global`, depending on how the name was defined in the parent scope. I don't really like that, though.

Reply:
That's pretty much how we do it in PEP 572, except of course it also has the special case that if there is neither `nonlocal x` nor `global x` in the containing scope, it becomes implicitly local in the containing scope (and non-local in the comprehension).